### PR TITLE
Security improvements: content-MD5

### PR DESCRIPTION
**Pull request by mike-podolskiy90**

### CODE REVIEW

File: src/main/java/org/gbif/ws/security/GbifAuthService.java [hunk 1]
```diff
@@ -124,8 +128,43 @@ private static String buildStringToSign(ContainerRequest request) {
...
+    final InputStream entityInputStream = request.getEntityInputStream();
+    Object entity = null;
+    try {
+      final int available = entityInputStream.available();
+
+      if (available > 0) {
+        entity = request.getEntity(Object.class);
+      }
+    } catch (IOException e) {
+      LOG.error("Can't get available bytes from the stream", e);
```
Comment (timrobertson100) [inline src/main/java/org/gbif/ws/security/GbifAuthService.java]: This looks suspicious. 

Later on `entity` is checked for `null` indicating that it is expected in some cases that an entity can't be read. Here though it is logged as an error. 

Is it an error (in which case shouldn't it throw Exception, or expected, in which case it should log at info.

File: src/main/java/org/gbif/ws/security/GbifAuthService.java [hunk 2]
```diff
@@ -240,9 +280,21 @@ public void signRequest(String username, ClientRequest request) {
...
+        entityMap = mapper.convertValue(entity, Map.class);
+      } else {
+        entityMap = ((Map) entity);
+      }
+
+      // filter null values
+      entityMap.values().removeIf(Objects::isNull);
+
+      byte[] content = mapper.writeValueAsBytes(entityMap);
       return new String(Base64.encode(DigestUtils.md5(content)), "ASCII");
```
Comment (fmendezh) [inline src/main/java/org/gbif/ws/security/GbifAuthService.java]: Does StandardCharsets.US_ASCII work?

File: src/main/java/org/gbif/ws/security/GbifAuthService.java [hunk 3]
```diff
@@ -175,8 +214,9 @@ private static String buildAuthHeader(String applicationKey, String signature) {
   /**
    * Generates a Base64 encoded HMAC-SHA1 signature of the passed in string with the given secret key.
    * See Message Authentication Code specs http://tools.ietf.org/html/rfc2104
+   *
    * @param stringToSign the string to be signed
-   * @param secretKey the secret key to use in the
+   * @param secretKey    the secret key to use in the
```
Comment (timrobertson100) [inline src/main/java/org/gbif/ws/security/GbifAuthService.java]: typo

File: src/main/java/org/gbif/ws/security/GbifAuthService.java [hunk 4]
```diff
@@ -124,8 +128,43 @@ private static String buildStringToSign(ContainerRequest request) {
       sb.append(getCanonicalizedPath(request.getRequestUri()));
     }
 
+    InputStream in = request.getEntityInputStream();
+    if (in.getClass() != ByteArrayInputStream.class) {
```
Comment (timrobertson100) [inline src/main/java/org/gbif/ws/security/GbifAuthService.java]: Please inline doc which this check

File: src/main/java/org/gbif/ws/security/GbifAuthService.java (hunk 2 above)
Comment (mike-podolskiy90) [reply src/main/java/org/gbif/ws/security/GbifAuthService.java]: just checked it's working. I'll change this

File: src/main/java/org/gbif/ws/security/GbifAuthService.java (hunk 3 above)
Comment (mike-podolskiy90) [reply src/main/java/org/gbif/ws/security/GbifAuthService.java]: fixed

File: src/main/java/org/gbif/ws/security/GbifAuthService.java (hunk 1 above)
Comment (mike-podolskiy90) [reply src/main/java/org/gbif/ws/security/GbifAuthService.java]: Exception can be thrown by InputStream#available if an unpredictable error happen. If the entity is null (no content was sent) then the stream should be empty and return 0

File: src/main/java/org/gbif/ws/security/GbifAuthService.java (hunk 4 above)
Comment (mike-podolskiy90) [reply src/main/java/org/gbif/ws/security/GbifAuthService.java]: done

File: src/main/java/org/gbif/ws/security/GbifAuthService.java (hunk 1 above)
Comment (timrobertson100) [reply src/main/java/org/gbif/ws/security/GbifAuthService.java]: Thanks - makes sense

But why only log then and not throw the Exception?

Comment (mike-podolskiy90) [reply src/main/java/org/gbif/ws/security/GbifAuthService.java]: Let's throw an RTE here

**Pull request merged**